core/src/main/java/org/apache/iceberg/avro/AvroFormatModel.java

```diff
@@ -0,0 +1,272 @@
+    public ReadBuilder<D, S> set(String key, String value) {
+      // Configuration is not used for Avro reader creation
+      return this;
+    }
+
+    @Override
+    public ReadBuilder<D, S> reuseContainers() {
+      internal.reuseContainers();
+      return this;
+    }
+
+    @Override
+    public ReadBuilder<D, S> recordsPerBatch(int numRowsPerBatch) {
+      throw new UnsupportedOperationException("Batch reading is not supported in Avro reader");
+    }
+
+    @Override
+    public ReadBuilder<D, S> idToConstant(Map<Integer, ?> newIdToConstant) {
+      this.idToConstant = newIdToConstant;
+      return this;
+    }
+
+    @Override
+    public ReadBuilder<D, S> withNameMapping(org.apache.iceberg.mapping.NameMapping nameMapping) {
+      internal.withNameMapping(nameMapping);
+      return this;
+    }
+
+    @Override
+    public ReadBuilder<D, S> withFileEncryptionKey(ByteBuffer encryptionKey) {
+      // Avro doesn't support file encryption via this method
+      throw new UnsupportedOperationException("Avro does not support file encryption keys");
+    }
+
+    @Override
+    public ReadBuilder<D, S> withAADPrefix(ByteBuffer aadPrefix) {
+      // Avro doesn't support AAD prefix
+      throw new UnsupportedOperationException("Avro does not support AAD prefix");
+    }
+
+    @Override
+    public CloseableIterable<D> build() {
+      return internal
+          .createResolvingReader(unused -> readerFunction.apply(icebergSchema, idToConstant))
+          .build();
+    }
+  }
+
+  private static class WriteBuilderWrapper<D, S> implements WriteBuilder<D, S> {
+    private final Avro.WriteBuilder internal;
+    private final BiFunction<org.apache.avro.Schema, S, DatumWriter<D>> writerFunction;
+    private S inputSchema;
+
+    private WriteBuilderWrapper(
+        EncryptedOutputFile outputFile,
+        BiFunction<org.apache.avro.Schema, S, DatumWriter<D>> writerFunction) {
+      this.internal = Avro.write(outputFile.encryptingOutputFile());
+      this.writerFunction = writerFunction;
+    }
+
+    @Override
+    public WriteBuilder<D, S> schema(Schema schema) {
+      internal.schema(schema);
+      return this;
+    }
+
+    @Override
+    public WriteBuilder<D, S> inputSchema(S schema) {
+      this.inputSchema = schema;
+      return this;
+    }
+
+    @Override
+    public WriteBuilder<D, S> set(String property, String value) {
+      internal.set(property, value);
+      return this;
+    }
+
+    @Override
+    public WriteBuilder<D, S> setAll(Map<String, String> properties) {
+      internal.setAll(properties);
+      return this;
+    }
+
+    @Override
+    public WriteBuilder<D, S> meta(String property, String value) {
+      internal.meta(property, value);
+      return this;
+    }
+
+    @Override
+    public WriteBuilder<D, S> meta(Map<String, String> properties) {
+      internal.meta(properties);
+      return this;
+    }
+
+    @Override
+    public WriteBuilder<D, S> content(FileContent content) {
```

Comment:
Configuration of the internal builder based on a combination of configuration calls (for instance, using both `content` and `inputSchema`) should happen in the `build` method. The purpose is to ensure that configuration is not based on the order of method calls and to avoid accidentally capturing state in closures.